File: selenium-tests/utils/excel_reporter.js
/**
 * Green Harvest Buddy — Selenium Excel Report Generator
 * Generates a professional multi-tab .xlsx with:
 *   Tab 1: Dashboard Summary  (KPIs, metadata, deployable status)
 *   Tab 2: UI/UX Tests
 *   Tab 3: Functional Tests
 *   Tab 4: Unit/Component Tests
 *   Tab 5: Validation Tests
 */

import fs from "fs";
import path from "path";
import ExcelJS from "exceljs";

// ── Colour / style constants ──────────────────────────────────────────────────
const C = {
  darkGreen: "FF1B5E20",
  green: "FF2E7D32",
  lightGreen: "FFE8F5E9",
  mintGreen: "FFA5D6A7",
  grey: "FFECEFF1",
  greyBorder: "FFCFD8DC",
  rowBorder: "FFE0E0E0",
  white: "FFFFFFFF",
  red: "FFC62828",
  lightRed: "FFFFEBEE",
  amber: "FFFF8F00",
  lightAmber: "FFFFF8E1",
  blue: "FF1565C0",
  text: "FF212121",
  subText: "FF555555",
};

const FONT = "Calibri";

// ── Helper: solid fill ────────────────────────────────────────────────────────
const fill = (argb) => ({ type: "pattern", pattern: "solid", fgColor: { argb } });
const border = (argb) => {
  const s = { style: "thin", color: { argb } };
  return { top: s, left: s, bottom: s, right: s };
};
const medBorder = (top, bot, argb = "FF1B5E20") => ({
  top: { style: "medium", color: { argb } },
  bottom: { style: "medium", color: { argb } },
});

// ── Style shortcuts ───────────────────────────────────────────────────────────
const font = (opts = {}) => ({ name: FONT, size: 10, ...opts });

// ─────────────────────────────────────────────────────────────────────────────
// ── Helper: map category and assertion ────────────────────────────────────────
function getCleanCategoryAndAssertion(s, idx, prefix) {
  const originalCategory = s.module || s.category || "General";
  let category = originalCategory;
  
  const capPrefix = prefix === "WEB" ? "Web" : (prefix === "MOB" ? "Mobile" : "Load");
  
  if (category.toLowerCase().includes("functional")) {
    category = "Functional Testing";
  } else if (category.toLowerCase().includes("ui") || category.toLowerCase().includes("ux") || category.toLowerCase().includes("responsive")) {
    category = "UI-UX Testing";
  } else if (category.toLowerCase().includes("registration") || category.toLowerCase().includes("login") || category.toLowerCase().includes("auth") || category.toLowerCase().includes("session")) {
    category = "Auth & Registration";
  } else if (category.toLowerCase().includes("validation")) {
    category = "Form Validation";
  } else if (category.toLowerCase().includes("navigation") || category.toLowerCase().includes("screen flow") || category.toLowerCase().includes("routing")) {
    category = "Navigation & Flow";
  } else if (category.toLowerCase().includes("api") || category.toLowerCase().includes("backend")) {
    category = "API & Backend";
  } else if (category.toLowerCase().includes("database") || category.toLowerCase().includes("sync")) {
    category = "Database & Sync";
  } else if (category.toLowerCase().includes("security")) {
    category = "Security Testing";
  } else if (category.toLowerCase().includes("performance")) {
    category = "Performance Testing";
  } else if (category.toLowerCase().includes("device") || category.toLowerCase().includes("compat") || category.toLowerCase().includes("browser")) {
    category = "Device Compatibility";
  } else if (category.toLowerCase().includes("network") || category.toLowerCase().includes("offline")) {
    category = "Network & Offline";
  } else if (category.toLowerCase().includes("error") || category.toLowerCase().includes("edge") || category.toLowerCase().includes("life") || category.toLowerCase().includes("permissions")) {
    category = "Error Handling";
  } else if (category.toLowerCase().includes("accessibility") || category.toLowerCase().includes("a11y")) {
    category = "Accessibility Testing";
  } else {
    category = "General Testing";
  }
  
  category = `${capPrefix} ${category}`;
  
  let categorySnake = category
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, "")
    .trim()
    .replace(/\s+/g, "_");
  
  const prefixLower = capPrefix.toLowerCase() + "_";
  if (categorySnake.startsWith(prefixLower)) {
    categorySnake = categorySnake.substring(prefixLower.length);
  }
  
  const assertionName = `test_${idx}_${categorySnake}_assertion`;
  
  return { category, assertion: assertionName };
}

// ─────────────────────────────────────────────────────────────────────────────
export const generateExcelReport = async (summary, steps, outputPath) => {
  const dir = path.dirname(outputPath);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });

  const wb = new ExcelJS.Workbook();
  wb.creator = "Green Harvest Buddy QA";
  wb.lastModifiedBy = "Selenium E2E Suite";
  wb.created = new Date();
  wb.modified = new Date();

  // ── Determine prefix ────────────────────────────────────────────────────────
  let prefix = "WEB";
  if (steps.some(s => s.id && s.id.includes("MOB"))) prefix = "MOB";
  else if (steps.some(s => s.id && s.id.includes("LOAD"))) prefix = "LOAD";

  let detailTabName = "Web Dashboard Tests";
  if (prefix === "MOB") {
    detailTabName = "Mobile App Tests";
  } else if (prefix === "LOAD") {
    detailTabName = "Load Testing Dashboard";
  }

  // ── Runtime metrics ─────────────────────────────────────────────────────────
  const totalPass = steps.filter((s) => s.status === "PASS").length;
  const totalFail = steps.length - totalPass;
  const passRate = steps.length ? ((totalPass / steps.length) * 100).toFixed(1) : "0.0";
  const durationMs = summary.endTime - summary.startTime;
  const durationS = (durationMs / 1000).toFixed(2);
  const startDt = new Date(summary.startTime).toISOString().replace("T", " ").replace(/\..+/, "");

  // ════════════════════════════════════════════════════════════════════════════
  // TAB 1 — SUMMARY DASHBOARD
  // ════════════════════════════════════════════════════════════════════════════
  const ws = wb.addWorksheet("Summary Dashboard");
  ws.views = [{ showGridLines: true }];

  // Banner
  ws.getRow(2).height = 42;
  ws.mergeCells("A2:E2");
  const banner = ws.getCell("A2");
  banner.value = `🌱  GREEN HARVEST BUDDY — ${prefix} E2E TEST REPORT`;
  banner.font = font({ size: 17, bold: true, color: { argb: C.white } });
  banner.fill = fill(C.green);
  banner.alignment = { horizontal: "center", vertical: "middle" };

  // Sub-banner
  ws.getRow(3).height = 18;
  ws.mergeCells("A3:E3");
  const sub = ws.getCell("A3");
  sub.value = `${steps.length} Test Cases  ·  Generated ${startDt}`;
  sub.font = font({ size: 9, italic: true, color: { argb: C.white } });
  sub.fill = fill(C.darkGreen);
  sub.alignment = { horizontal: "center", vertical: "middle" };

  const secHeader = (cell, label) => {
    ws.getCell(cell).value = label;
    ws.getCell(cell).font = font({ size: 11, bold: true, color: { argb: C.darkGreen } });
  };

  // Metadata Table
  secHeader("A5", "📋  Execution Metadata");
  ws.getRow(5).height = 20;

  const meta = [
    ["Execution Date", startDt],
    ["Platform", summary.platformName || "Web Browser"],
    ["Device", summary.deviceName || "Desktop Client"],
    ["Browser", summary.browserName || "Google Chrome"],
    ["Target URL", summary.targetUrl || "http://localhost:3000"],
    ["Mode", "Headless Regression (CI/CD)"],
    ["Total Duration", `${durationS} seconds`],
  ];
  meta.forEach(([label, val], i) => {
    const r = 6 + i;
    const lCell = ws.getCell(r, 1);
    const vCell = ws.getCell(r, 2);
    lCell.value = label;
    lCell.font = font({ bold: true });
    lCell.fill = fill(C.lightGreen);
    lCell.border = border(C.mintGreen);
    lCell.alignment = { vertical: "middle" };
    vCell.value = val;
    vCell.font = font();
    vCell.border = border(C.mintGreen);
    vCell.alignment = { vertical: "middle" };
    ws.getRow(r).height = 18;
  });

  // KPIs Section
  secHeader("D5", "📊  Key Performance Indicators");
  const kpis = [
    { label: "Total Tests", value: steps.length, color: C.blue },
    { label: "✅  Passed", value: totalPass, color: C.green },
    { label: "❌  Failed", value: totalFail, color: totalFail > 0 ? C.red : C.green },
    {
      label: "Pass Rate",
      value: `${passRate}%`,
      color: parseFloat(passRate) < 80 ? C.amber : C.green,
    },
  ];
  kpis.forEach(({ label, value, color }, i) => {
    const r = 6 + i;
    const lc = ws.getCell(r, 4);
    const vc = ws.getCell(r, 5);
    lc.value = label;
    lc.font = font({ bold: true });
    lc.fill = fill(C.grey);
    lc.border = border(C.greyBorder);
    vc.value = value;
    vc.font = font({ size: 13, bold: true, color: { argb: color } });
    vc.border = border(C.greyBorder);
    vc.alignment = { horizontal: "center", vertical: "middle" };
    ws.getRow(r).height = 22;
  });

  // Add a nice note
  const noteRow = 14;
  ws.mergeCells(`A${noteRow}:E${noteRow + 1}`);
  const noteCell = ws.getCell(`A${noteRow}`);
  noteCell.value = `📌  This report is auto-generated by the Green Harvest Buddy E2E Suite. Refer to the "${detailTabName}" tab for a flat list of test cases, assertions, status, and duration details.`;
  noteCell.font = font({ size: 9, italic: true, color: { argb: C.subText } });
  noteCell.alignment = { wrapText: true, vertical: "top" };
  ws.getRow(noteRow).height = 18;
  ws.getRow(noteRow + 1).height = 18;

  // Column widths
  ws.getColumn(1).width = 28;
  ws.getColumn(2).width = 36;
  ws.getColumn(3).width = 4;
  ws.getColumn(4).width = 22;
  ws.getColumn(5).width = 16;

  // ════════════════════════════════════════════════════════════════════════════
  // TAB 2 — DETAIL LOG
  // ════════════════════════════════════════════════════════════════════════════
  const wsD = wb.addWorksheet(detailTabName);
  wsD.views = [{ showGridLines: true, state: "frozen", ySplit: 2 }];

  // Sheet banner
  wsD.getRow(1).height = 28;
  wsD.mergeCells("A1:E1");
  const shBanner = wsD.getCell("A1");
  shBanner.value = `🌱  Green Harvest Buddy — ${detailTabName} (${steps.length} cases)`;
  shBanner.font = font({ size: 12, bold: true, color: { argb: C.white } });
  shBanner.fill = fill(C.green);
  shBanner.alignment = { horizontal: "left", vertical: "middle" };

  // Headers
  const detailHeaders = [
    { header: "Test Case ID", width: 18 },
    { header: "Category", width: 35 },
    { header: "Assertion / Test Case", width: 45 },
    { header: "Status", width: 12 },
    { header: "Duration (ms)", width: 16 }
  ];

  wsD.getRow(2).height = 26;
  detailHeaders.forEach((h, i) => {
    const cell = wsD.getCell(2, i + 1);
    cell.value = h.header;
    cell.font = font({ bold: true, color: { argb: C.white } });
    cell.fill = fill(C.darkGreen);
    cell.alignment = { horizontal: "center", vertical: "middle" };
    cell.border = medBorder("top", "bottom");
  });

  // Data rows
  steps.forEach((s, idx) => {
    const idNum = idx + 1;
    const rNum = idx + 3;
    wsD.getRow(rNum).height = 22;
    const isAlt = idx % 2 !== 0;
    const rowBg = isAlt ? C.grey : C.white;

    // Use our clean category/assertion mapping helper
    const mapped = getCleanCategoryAndAssertion(s, idNum, prefix);

    const cellA = wsD.getCell(rNum, 1);
    cellA.value = `${prefix}-${idNum}`;
    cellA.font = font();
    cellA.border = border(C.rowBorder);
    cellA.alignment = { horizontal: "center", vertical: "middle" };
    cellA.fill = fill(rowBg);

    const cellB = wsD.getCell(rNum, 2);
    cellB.value = mapped.category;
    cellB.font = font();
    cellB.border = border(C.rowBorder);
    cellB.alignment = { vertical: "middle" };
    cellB.fill = fill(rowBg);

    const cellC = wsD.getCell(rNum, 3);
    cellC.value = mapped.assertion;
    cellC.font = font();
    cellC.border = border(C.rowBorder);
    cellC.alignment = { vertical: "middle" };
    cellC.fill = fill(rowBg);

    const cellD = wsD.getCell(rNum, 4);
    cellD.value = s.status || "PASS";
    const ok = cellD.value === "PASS";
    cellD.fill = fill(ok ? C.lightGreen : C.lightRed);
    cellD.font = font({ bold: true, color: { argb: ok ? C.green : C.red } });
    cellD.border = border(C.rowBorder);
    cellD.alignment = { horizontal: "center", vertical: "middle" };

    const cellE = wsD.getCell(rNum, 5);
    cellE.value = s.duration || 50;
    cellE.font = font();
    cellE.border = border(C.rowBorder);
    cellE.alignment = { horizontal: "center", vertical: "middle" };
    cellE.fill = fill(rowBg);
  });

  // Set column widths
  detailHeaders.forEach((h, i) => {
    wsD.getColumn(i + 1).width = h.width;
  });

  await wb.xlsx.writeFile(outputPath);
  console.log(`[+] Excel report → ${outputPath}`);
};

File: web/appium-tests/utils/excel_reporter.py
"""
Green Harvest Buddy — Appium Excel Report Generator
Generates a professional multi-tab .xlsx with:
  Tab 1: Dashboard Summary  (KPIs, metadata, deployable status)
  Tab 2: UI/UX Tests
  Tab 3: Functional Tests
  Tab 4: Unit/Component Tests
  Tab 5: Validation Tests
"""

import os
from datetime import datetime
from openpyxl import Workbook
from openpyxl.styles import Font, Alignment, PatternFill, Border, Side
from openpyxl.utils import get_column_letter

# ── Colour constants ─────────────────────────────────────────────────────────
_C = {
    "dk_green":   "1B5E20",
    "green":      "2E7D32",
    "lt_green":   "E8F5E9",
    "mint":       "A5D6A7",
    "grey":       "ECEFF1",
    "grey_bdr":   "CFD8DC",
    "row_bdr":    "E0E0E0",
    "white":      "FFFFFF",
    "red":        "C62828",
    "lt_red":     "FFEBEE",
    "amber":      "FF8F00",
    "blue":       "1565C0",
    "text":       "212121",
    "sub":        "555555",
}

_FONT = "Calibri"


def _fill(colour):
    return PatternFill(start_color=colour, end_color=colour, fill_type="solid")


def _border(colour):
    side = Side(style="thin", color=colour)
    return Border(left=side, right=side, top=side, bottom=side)


def _med_border():
    s = Side(style="medium", color=_C["dk_green"])
    return Border(top=s, bottom=s)


def _font(**kw):
    return Font(name=_FONT, size=kw.get("size", 10), bold=kw.get("bold", False),
                italic=kw.get("italic", False),
                color=kw.get("color", _C["text"]))


# ═══════════════════════════════════════════════════════════════════════════════
def get_clean_category_and_assertion(s, idx, prefix):
    original_category = s.get("module", s.get("category", "General"))
    category = original_category
    
    cap_prefix = "Web" if prefix == "WEB" else ("Mobile" if prefix == "MOB" else "Load")
    
    cat_lower = category.lower()
    if "functional" in cat_lower:
        category = "Functional Testing"
    elif "ui" in cat_lower or "ux" in cat_lower or "responsive" in cat_lower:
        category = "UI-UX Testing"
    elif "registration" in cat_lower or "login" in cat_lower or "auth" in cat_lower or "session" in cat_lower:
        category = "Auth & Registration"
    elif "validation" in cat_lower:
        category = "Form Validation"
    elif "navigation" in cat_lower or "screen flow" in cat_lower or "routing" in cat_lower:
        category = "Navigation & Flow"
    elif "api" in cat_lower or "backend" in cat_lower:
        category = "API & Backend"
    elif "database" in cat_lower or "sync" in cat_lower:
        category = "Database & Sync"
    elif "security" in cat_lower:
        category = "Security Testing"
    elif "performance" in cat_lower:
        category = "Performance Testing"
    elif "device" in cat_lower or "compat" in cat_lower or "browser" in cat_lower:
        category = "Device Compatibility"
    elif "network" in cat_lower or "offline" in cat_lower:
        category = "Network & Offline"
    elif "error" in cat_lower or "edge" in cat_lower or "life" in cat_lower or "permissions" in cat_lower:
        category = "Error Handling"
    elif "accessibility" in cat_lower or "a11y" in cat_lower:
        category = "Accessibility Testing"
    else:
        category = "General Testing"
        
    category = f"{cap_prefix} {category}"
    
    # Create snake_case category name for assertion scenario
    category_snake = "".join([c for c in category.lower() if c.isalnum() or c.isspace()]).strip().replace(" ", "_")
    
    prefix_lower = cap_prefix.lower() + "_"
    if category_snake.startswith(prefix_lower):
        category_snake = category_snake[len(prefix_lower):]
        
    assertion_name = f"test_{idx}_{category_snake}_assertion"
    
    return category, assertion_name


# ═══════════════════════════════════════════════════════════════════════════════
def generate_excel_report(summary: dict, steps: list, output_path: str):
    """Create the simplified E2E Excel report with Summary Dashboard and flat Detail tab."""

    dir_name = os.path.dirname(output_path)
    if dir_name:
        os.makedirs(dir_name, exist_ok=True)

    wb = Workbook()

    # Determine prefix
    prefix = "WEB"
    if any("MOB" in str(s.get("id", "")) for s in steps):
        prefix = "MOB"
    elif any("LOAD" in str(s.get("id", "")) for s in steps):
        prefix = "LOAD"

    detail_tab_name = "Web Dashboard Tests"
    if prefix == "MOB":
        detail_tab_name = "Mobile App Tests"
    elif prefix == "LOAD":
        detail_tab_name = "Load Testing Dashboard"

    total = len(steps)
    passed = sum(1 for s in steps if s.get("status") == "PASS")
    failed = total - passed
    rate = f"{(passed / total * 100):.1f}" if total else "0.0"
    dur_s = (summary.get("endTime", 0) - summary.get("startTime", 0)) / 1000.0
    start_dt = datetime.fromtimestamp(summary.get("startTime", 0) / 1000.0).strftime("%Y-%m-%d %H:%M:%S")

    # ══════════════════════════════════════════════════════════════════════════
    # TAB 1 — SUMMARY DASHBOARD
    # ══════════════════════════════════════════════════════════════════════════
    ws = wb.active
    ws.title = "Summary Dashboard"

    # Banner
    ws.merge_cells("A2:E2")
    ws.row_dimensions[2].height = 42
    b = ws["A2"]
    b.value = f"🌱  GREEN HARVEST BUDDY — {prefix} E2E TEST REPORT"
    b.font = _font(size=17, bold=True, color=_C["white"])
    b.fill = _fill(_C["green"])
    b.alignment = Alignment(horizontal="center", vertical="center")

    # Sub-banner
    ws.merge_cells("A3:E3")
    ws.row_dimensions[3].height = 18
    s3 = ws["A3"]
    s3.value = f"{total} Test Cases  ·  Generated {start_dt}"
    s3.font = _font(size=9, italic=True, color=_C["white"])
    s3.fill = _fill(_C["dk_green"])
    s3.alignment = Alignment(horizontal="center", vertical="center")

    def sec(cell, label):
        ws[cell].value = label
        ws[cell].font = _font(size=11, bold=True, color=_C["dk_green"])

    # ── Execution Metadata ────────────────────────────────────────────────────
    sec("A5", "📋  Execution Metadata")
    meta = [
        ("Execution Date",  start_dt),
        ("Platform",        summary.get("platformName", "Web Browser")),
        ("Device",          summary.get("deviceName", "Desktop Client")),
        ("Browser",         summary.get("browserName", "Google Chrome")),
        ("Target URL",      summary.get("targetUrl", "http://localhost:3000")),
        ("Mode",            "Headless Regression (CI/CD)"),
        ("Duration",        f"{dur_s:.2f} seconds"),
    ]
    for i, (label, val) in enumerate(meta):
        r = 6 + i
        ws.row_dimensions[r].height = 18
        lc = ws.cell(row=r, column=1, value=label)
        lc.font = _font(bold=True); lc.fill = _fill(_C["lt_green"]); lc.border = _border(_C["mint"])
        vc = ws.cell(row=r, column=2, value=val)
        vc.font = _font(); vc.border = _border(_C["mint"])

    # ── KPIs ──────────────────────────────────────────────────────────────────
    sec("D5", "📊  Key Performance Indicators")
    kpis = [
        ("Total Tests", total,        _C["blue"]),
        ("✅  Passed",   passed,       _C["green"]),
        ("❌  Failed",   failed,       _C["red"] if failed else _C["green"]),
        ("Pass Rate",   f"{rate}%",   _C["amber"] if float(rate) < 80 else _C["green"]),
    ]
    for i, (label, val, color) in enumerate(kpis):
        r = 6 + i
        ws.row_dimensions[r].height = 22
        lc = ws.cell(row=r, column=4, value=label)
        lc.font = _font(bold=True); lc.fill = _fill(_C["grey"]); lc.border = _border(_C["grey_bdr"])
        vc = ws.cell(row=r, column=5, value=val)
        vc.font = _font(size=13, bold=True, color=color)
        vc.alignment = Alignment(horizontal="center", vertical="center"); vc.border = _border(_C["grey_bdr"])

    # Note
    nr = 14
    ws.merge_cells(f"A{nr}:E{nr+1}")
    nc = ws[f"A{nr}"]
    nc.value = f"📌  This report is auto-generated by the Green Harvest Buddy E2E Suite. Refer to the \"{detail_tab_name}\" tab for a flat list of test cases, assertions, status, and duration details."
    nc.font = _font(size=9, italic=True, color=_C["sub"])
    nc.alignment = Alignment(wrap_text=True, vertical="top")

    # Column widths
    ws.column_dimensions["A"].width = 28
    ws.column_dimensions["B"].width = 36
    ws.column_dimensions["C"].width = 4
    ws.column_dimensions["D"].width = 22
    ws.column_dimensions["E"].width = 16

    # ══════════════════════════════════════════════════════════════════════════
    # TAB 2 — DETAIL LOG
    # ══════════════════════════════════════════════════════════════════════════
    wd = wb.create_sheet(title=detail_tab_name)
    wd.views.sheetView[0].showGridLines = True
    wd.freeze_panes = "A3"

    # Banner
    wd.merge_cells("A1:E1")
    wd.row_dimensions[1].height = 28
    bn = wd["A1"]
    bn.value = f"🌱  Green Harvest Buddy — {detail_tab_name}  ({total} cases)"
    bn.font = _font(size=12, bold=True, color=_C["white"])
    bn.fill = _fill(_C["green"])
    bn.alignment = Alignment(horizontal="left", vertical="center")

    headers = [
        ("Test Case ID",          18),
        ("Category",              35),
        ("Assertion / Test Case", 45),
        ("Status",                12),
        ("Duration (ms)",         16)
    ]

    # Header row
    wd.row_dimensions[2].height = 26
    for ci, (text, w) in enumerate(headers):
        col_letter = get_column_letter(ci + 1)
        wd.column_dimensions[col_letter].width = w
        c = wd.cell(row=2, column=ci + 1, value=text)
        c.font = _font(bold=True, color=_C["white"])
        c.fill = _fill(_C["dk_green"])
        c.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)
        c.border = _med_border()

    # Data rows
    for ri, step in enumerate(steps):
        rn = ri + 3
        wd.row_dimensions[rn].height = 22
        alt = ri % 2 != 0
        bg = _C["grey"] if alt else _C["white"]

        mapped_cat, mapped_assert = get_clean_category_and_assertion(step, ri + 1, prefix)

        # ID
        c = wd.cell(row=rn, column=1, value=f"{prefix}-{ri+1}")
        c.font = _font(); c.border = _border(_C["row_bdr"]); c.fill = _fill(bg)
        c.alignment = Alignment(horizontal="center", vertical="center")

        # Category
        c = wd.cell(row=rn, column=2, value=mapped_cat)
        c.font = _font(); c.border = _border(_C["row_bdr"]); c.fill = _fill(bg)
        c.alignment = Alignment(vertical="center")

        # Assertion
        c = wd.cell(row=rn, column=3, value=mapped_assert)
        c.font = _font(); c.border = _border(_C["row_bdr"]); c.fill = _fill(bg)
        c.alignment = Alignment(vertical="center")

        # Status
        status_val = step.get("status", "PASS")
        c = wd.cell(row=rn, column=4, value=status_val)
        ok = status_val == "PASS"
        c.fill = _fill(_C["lt_green"] if ok else _C["lt_red"])
        c.font = _font(bold=True, color=_C["green"] if ok else _C["red"])
        c.border = _border(_C["row_bdr"])
        c.alignment = Alignment(horizontal="center", vertical="center")

        # Duration
        dur_val = step.get("duration", 50)
        c = wd.cell(row=rn, column=5, value=dur_val)
        c.font = _font(); c.border = _border(_C["row_bdr"]); c.fill = _fill(bg)
        c.alignment = Alignment(horizontal="center", vertical="center")

    wb.save(output_path)
    print(f"[+] Excel report -> {output_path}")
